### PR TITLE
Hiding vtables to enable more compiler optimizations.

### PR DESCRIPTION
**Pull request by benvanik**
Progress on #6641.

### CODE REVIEW

File: iree/hal/vulkan/native_event.cc [hunk 1]
```diff
@@ -22,7 +22,9 @@ typedef struct iree_hal_vulkan_native_event_t {
   VkEvent handle;
 } iree_hal_vulkan_native_event_t;
 
+namespace {
 extern const iree_hal_event_vtable_t iree_hal_vulkan_native_event_vtable;
```
Comment (stellaraccident) [inline iree/hal/vulkan/native_event.cc:26]: I'm not sure what this is trying to do (extern in an anonymous namespace)

File: iree/hal/vulkan/native_executable.cc [hunk 2]
```diff
@@ -199,8 +199,10 @@ typedef struct iree_hal_vulkan_native_executable_t {
   iree_hal_vulkan_entry_point_t entry_points[];
 } iree_hal_vulkan_native_executable_t;
 
+namespace {
```
Comment (stellaraccident) [inline iree/hal/vulkan/native_executable.cc:202]: Ditto

**Review by stellaraccident — APPROVED** [review]
One question.

File: iree/hal/vulkan/native_event.cc (hunk 1 above)
Comment (benvanik) [reply iree/hal/vulkan/native_event.cc:26]: In C++ the `extern` is required to forward declare the value that's defined below and the anonymous namespace keeps the whole thing local. Issue is that in C the same identifier used within a compilation unit refers to the same instance (so two `static int foo` will reference the same memory, only one of which may contain a definition), while in C++ they don't and so the extern is effectively telling the compiler not to unique it. It's ... not a stellar feature of the language :P

**Comment by GMNGeoffrey** [comment]
Looks like this is breaking cuda tests

> iree-check-module: ../iree/hal/device.c:32: iree_hal_allocator_t *iree_hal_device_allocator(iree_hal_device_t *): Assertion `device' failed.
/tmpfs/src/github/iree/build_tools/cmake/run_test.sh: line 25: 30150 Aborted                 (core dumped) "$@"

https://source.cloud.google.com/results/invocations/e5b6feda-c5e9-4ff2-a8e2-b7777cbb4f15/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-turing%2Fmain/log

**Pull request merged**